Step 3 - 자동차 경주

ohsung opened the pull request:
코드 리뷰 요청 드립니다.

감사합니다.

ohsung committed:
feat : 자동차 경주

step 2 수정사항

1. 문자열 계산기 Main 입출력 수정
2. build.gradle.kts kotest 5.5.1 수정
3. Operator operation 고차함수 수정
4. execute함수에서 validation책임을 분리 - requireStr 함수를 추가 하였는데 책임분리가 맞나요?
5. 피연산자 정규식 표현을 위해 toCalInt 추가

namjackson commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -5,18 +5,23 @@ object Calculator {
     }
 
+    private fun requireStr(str: String?): String {
+        require(!str.isNullOrEmpty()) {
+            "입력값이 null이거나 빈 공백 문자일 경우"
+        }
```
문자열의 검증에 대한 객체를 따로 분리해보면 좋을거같아요!
문자열에 대한 테스트코드도 간단하게 테스트할수 있을거같단 생각이 들어요!

namjackson commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -1,20 +1,12 @@
+    DIVIDE("/", { x, y -> x / y });
 
-    abstract fun operation(x: Int, y: Int): Int
+    // abstract fun operation(x: Int, y: Int): Int
```
불필요해진 코드는 과감히 제거해주세요 👍

namjackson commented on src/main/kotlin/race/model/Race.kt:
```diff
@@ -0,0 +1,30 @@
+    private val numberOfCar: Int,
+) {
+
+    companion object {
+        const val MOVING_NUMBER = 4
+    }
```
companion object 는 클래스의 최하단에 위치해주시면 좋을거같아요!
아래 컨벤션을 확인해주셔도 좋을거같네요!
https://kotlinlang.org/docs/coding-conventions.html#class-layout

namjackson commented on src/main/kotlin/race/model/Race.kt:
```diff
@@ -0,0 +1,30 @@
+        }
+    }
+
+    val cars: List<Car> get() = _cars
```
cars로 받은 carList를 수정하게되면 
_cars도 영향을 받는 위험한 구조네요!
방어적 복사를 고민해보셔도 좋을거같아요!

namjackson commented on src/main/kotlin/race/model/Race.kt:
```diff
@@ -0,0 +1,30 @@
+        }
+    }
+
+    fun go(car: Car = _cars[0], number: Int = (0..9).random()) {
+        if (number >= MOVING_NUMBER) {
+            car.move()
+        }
+    }
```
go함수는 테스트를 제외하면 race내부에서만 호출되고 있어요, 
private으로 변경해주세요!
테스트가능한 구조는 중요하지만, 테스트만을 위한 코드는 지양해주세요!

car, number 파라미터는 테스트를 위한 파라미터 아닐까요?
_cars[0], (0..9).random() 이 무엇을 의미하는지 파악하기가 힘든거 같아요!

namjackson replied:
> 랜덤 테스트 코드가 어렵네요..

성오님 말씀대로, 랜덤관련 기능은 테스트하기가 힘들곤합니다! 
테스트를 위해 랜덤 관련 코드에 대한 의존성을 분리해보세요! 
(랜덤관련 기능을 클래스로 분리해보세요)

namjackson commented on src/main/kotlin/study/main.kt:
```diff
@@ -3,5 +3,8 @@ package study
 
 fun main() {
-    Calculator.execute("5 + - 1 * 6 / 2")
+    val expression = "2 + 3 * 4 / 2"
+    println("입력 $expression")
```
expression을 콘솔에서 입력받도록 해주세요 😄

namjackson reviewed (changes requested):
안녕하세요 성오님 ! 
3단계 잘 진행해주셨네요! 
추가로 고민할법한 포인트들 코멘트로 남겼으니, 확인해주세요! 
피드백 적용후, 다시 리뷰요청해주시면 감사하겠습니다!

ohsung committed:
refactor : 자동차 경주, 문자열 계산기

1. CalculatorValidate 유효성 체크
2. 방어적 복사 .toList() 처리
3. random 처리를 위한 Engine interface 추가

namjackson commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -5,24 +5,21 @@ object Calculator {
+        val args = s.split(DELIMITER)
 
-        val nums = args.filterIndexed { index, _ -> index.rem(2) == 0 }.map { it.toInt() }
+        val nums = args.filterIndexed { index, _ -> index.rem(2) == 0 }.map { CalculatorValidate.number(it) }
```
CalculatorValidate. 처럼 싱글턴으로 접근하는 방법은, 
편한지만, 객체간 결합도를 올리곤해요,
생성시점에 주입받아서사용해보면 어떨까요?

namjackson commented on src/main/kotlin/race/controller/RacingController.kt:
```diff
@@ -0,0 +1,26 @@
+        resultView.showResult()
+        // resultView.showRacing(racing.cars)
+
+        repeat(numberOfRace) {
+            race.carsMove()
```
numberOfRace만큼 레이스를 진행시키는건 Controller가 아닌 Race객체의 책임은 아닐까요?
구현에는 정답이 없으니, 생각해보셔도 좋을거같아요

namjackson commented on src/main/kotlin/race/model/Engine.kt:
```diff
@@ -0,0 +1,5 @@
+package race.model
+
+interface Engine {
+    fun create(): Int
```
engine 추상화 👍  
하지만 create이라는 함수는 무엇을 의미할까요?
engine은 숫자 생성이 아닌, 이동에 대한 추상화 아닐까요?

움직일지 말지에 대한 부분이 Engine의 관심사인거같아요!

namjackson commented on src/main/kotlin/race/model/Race.kt:
```diff
@@ -0,0 +1,20 @@
+        }
+    }
+
+    val cars: List<Car> get() = _cars.toList()
```
방어적 복사 👍

namjackson commented on src/test/kotlin/race/RacingCarTest.kt:
```diff
@@ -0,0 +1,30 @@
+import race.model.Race
+
+class RacingCarTest : StringSpec({
+    "Car 움직임 2번 테스트" {
```
테스트케이스명도 구체적으로 작성해보면 어덜까요? 
저는 케이스명의 경우도, given, when, then 표기법을 활용하곤 해요!
`항상 이동하는 엔진으로, Car가 2번 이동하면, Car의 위치는 2이다.`

namjackson commented on src/test/kotlin/race/RacingCarTest.kt:
```diff
@@ -0,0 +1,30 @@
+        race.cars.size shouldBe 5
+    }
+
+    "Car power 4 이상일 경우 움직임" {
+        val lowCar = Car(engine = EngineTest(1))
+        lowCar.go()
+        lowCar.position shouldBe 0
+
```
테스트명과 테스트 케이스가 일치하지 않아요!

namjackson commented on src/test/kotlin/race/EngineTest.kt:
```diff
@@ -0,0 +1,7 @@
+
+import race.model.Engine
+
+class EngineTest(private val power: Int) : Engine {
```
EngineTest보다는 FakeEngine 같은 네이밍도 좋은거같아요!

namjackson commented on src/test/kotlin/race/RacingCarTest.kt:
```diff
@@ -0,0 +1,30 @@
+import race.model.Car
+import race.model.Race
+
+class RacingCarTest : StringSpec({
```
Car에 대한 테스트만 있네요!
Race 클래스의 테스트코드도 작성해보면 좋을거 같아요!

namjackson commented on README.md:
```diff
@@ -9,3 +9,12 @@
++ 주어진 횟수 동안 n대의 자동차는 전진 또는 멈출 수 있다.
++ 사용자는 몇 대의 자동차로 몇 번의 이동을 할 것인지를 입력할 수 있어야 한다.
++ 전진하는 조건은 0에서 9 사이에서 무작위 값을 구한 후 무작위 값이 4 이상일 경우이다.
++ 자동차의 상태를 화면에 출력한다. 어느 시점에 출력할 것인지에 대한 제약은 없다.
```
커밋을 좀 더 작게 나누면 좋을거 같아요

namjackson reviewed (approved):
안녕하세요 성오님,
피드백 적용 잘해주셨네요! 👍 
추가적인 코멘트 남겼으니, 확인해주세요! 
변경사항은 다음 미션 진행하면서 적용해주세요!!
